File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int maxProfit = 0;
        int n = prices.size();
        int bestBuy = prices[0];
        int index = 0;

        for(int i = 0; i < n; i++){
            if(bestBuy > prices[i]){
              bestBuy = prices[i];
              index = i;
            }
        }

        for(int i=index+1; i<n; i++ ){
            int profit = prices[i] - bestBuy;
            maxProfit = max(maxProfit, profit);
        }

        return maxProfit;
    }
};

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
// Lets more optimize this
function maxProfit(prices) {

 let profits = 0;
 let bestBuy = prices[0];

 for(let price of prices)
 {
    bestBuy = Math.min(bestBuy, price);
    if(price - bestBuy > 0){
        profits = Math.max(profits, price - bestBuy);
    }
 }

 return profits;
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
//   let minStockPurchasePrice = prices[0] || 0;
//   let profits = 0;

//   for (let i = 1; i < prices.length; i++) {
//     if (prices[i] < minStockPurchasePrice) {
//       minStockPurchasePrice = prices[i];
//     }
//     profits = Math.max(profits,  prices[i] - minStockPurchasePrice);
//   }

//   return profits;
}